scripts/validate_docs_snippet_line_numbers.py

```diff
@@ -0,0 +1,202 @@
+            "Checking %s for snippet validity",
+            docs_file,
+        )
+        file_broken_refs: List[DocusaurusRef] = _evaluate_snippet_validity(
+            docs_file, ref_list
+        )
+        all_broken_refs.extend(file_broken_refs)
+
+    return all_broken_refs
+
+
+def _evaluate_snippet_validity(
+    file: str, ref_list: List[DocusaurusRef]
+) -> List[DocusaurusRef]:
+    with open(file) as f:
+        file_contents: List[str] = f.readlines()
+
+    broken_refs: List[DocusaurusRef] = []
+    for ref in ref_list:
+        if not _validate_docusaurus_ref(ref, file_contents):
+            broken_refs.append(ref)
+
+    return broken_refs
+
+
+def _validate_docusaurus_ref(
+    docusaurus_ref: DocusaurusRef, file_contents: List[str]
+) -> bool:
+    start_line, end_line = docusaurus_ref.line_numbers
+    if start_line > len(file_contents) or end_line > len(file_contents):
+        logger.warning(
+            "Invalid snippet due to out-of-bounds reference: %s", docusaurus_ref
+        )
+        return False
+
+    return _validate_opening_tag(start_line, file_contents) and _validate_closing_tag(
+        end_line, file_contents
+    )
+
+
+def _validate_opening_tag(start_line: int, file_contents: List[str]) -> bool:
+    ptr: int = start_line - 1
+
+    line: str
+    while ptr >= 0:
+        line = file_contents[ptr - 1]
+        if OPENING_TAG in line:
+            return True
+        elif not line.isspace():
+            return False
+        ptr -= 1
+
+    return False
+
+
+def _validate_closing_tag(end_line: int, file_contents: List[str]) -> bool:
+    ptr: int = end_line + 1
+
+    line: str
+    while ptr < len(file_contents):
+        line = file_contents[ptr - 1]
+        if CLOSING_TAG in line:
+            return True
+        elif not line.isspace():
+            return False
+        ptr += 1
+
+    return False
+
+
+def print_diagnostic_report(broken_refs: List[DocusaurusRef]) -> None:
+    """Prints results to console for the user to review.
+
+    Args:
+        broken_refs: A list of broken refernces (as parsed and determined by prior stages).
+    """
+    refs_by_source_doc: Dict[str, List[DocusaurusRef]] = defaultdict(list)
+    for ref in broken_refs:
+        refs_by_source_doc[ref.docs_file].append(ref)
+
+    for docs_file, refs in refs_by_source_doc.items():
+        print(f"\n{docs_file}:")
+        for ref in refs:
+            start, end = ref.line_numbers
+            print(f"  {ref.py_file}: L{start}-L{end}")
+
+    print(
+        f"\n[ERROR] {len(broken_refs)} snippet issue(s) found in {len(refs_by_source_doc)} docs file(s)!"
+    )
+
+
+if __name__ == "__main__":
+    docusaurus_refs: Dict[str, List[DocusaurusRef]] = collect_docusaurus_refs("docs")
+    broken_refs: List[DocusaurusRef] = evaluate_snippet_validity(docusaurus_refs)
+    if broken_refs:
+        print_diagnostic_report(broken_refs)
+        # sys.exit(1) # TODO(cdkini): Enable once all errors are resolved
```

Comment:
This will be flipped on once we want to make this script blocking (after we fix all the docs)